bundles/org.openhab.core/src/main/java/org/openhab/core/cache/ByteArrayFileCache.java

```diff
@@ -0,0 +1,304 @@
+    private static final char EXTENSION_SEPARATOR = '.';
+
+    private final File cacheFolder;
+
+    static final long ONE_DAY_IN_MILLIS = TimeUnit.DAYS.toMillis(1);
+    private long expiry;
+
+    private static final Map<String, File> FILES_IN_CACHE = new ConcurrentHashMap<>();
+
+    /**
+     * Creates a new {@link ByteArrayFileCache} instance for a service. Creates a <code>cache</code> folder under
+     * <code>$OPENHAB_USERDATA/cache/$servicePID</code>.
+     *
+     * @param servicePID PID of the service
+     */
+    public ByteArrayFileCache(String servicePID) {
+        // TODO track and limit folder size
+        // TODO support user specific folder
+        cacheFolder = new File(new File(OpenHAB.getUserDataFolder(), CACHE_FOLDER_NAME), servicePID);
+        if (!cacheFolder.exists()) {
+            logger.debug("Creating cache folder '{}'", cacheFolder.getAbsolutePath());
+            cacheFolder.mkdirs();
+        }
+        logger.debug("Using cache folder '{}'", cacheFolder.getAbsolutePath());
+    }
+
+    /**
+     * Creates a new {@link ByteArrayFileCache} instance for a service. Creates a <code>cache</code> folder under
+     * <code>$OPENHAB_USERDATA/cache/$servicePID/</code>.
+     *
+     * @param servicePID PID of the service
+     * @param long the days for how long the files stay in the cache valid. Must be positive. 0 to
+     *            disable this functionality.
+     */
+    public ByteArrayFileCache(String servicePID, long expiry) {
+        this(servicePID, Duration.ofDays(expiry));
+    }
+
+    /**
+     * Creates a new {@link ByteArrayFileCache} instance for a service. Creates a <code>cache</code> folder under
+     * <code>$OPENHAB_USERDATA/cache/$servicePID/</code>.
+     *
+     * @param servicePID PID of the service
+     * @param Duration the days for how long the files stay in the cache valid. Must be positive. 0 to
+     *            disable this functionality.
+     */
+    public ByteArrayFileCache(String servicePID, Duration expiry) {
+        this(servicePID);
+        if (expiry.isNegative()) {
+            throw new IllegalArgumentException("Cache expiration time must be greater than or equal to 0");
+        }
+        this.expiry = expiry.toDays();
+    }
+
+    /**
+     * Adds a file to the cache. If the cache previously contained a file for the key, the old file is replaced by the
+     * new content.
+     *
+     * @param key the key with which the file is to be associated
+     * @param content the content for the file to be associated with the specified key
+     */
+    public void put(String key, byte[] content) {
+        writeFile(getUniqueFile(key), content);
+    }
+
+    /**
+     * Adds a file to the cache.
+     *
+     * @param key the key with which the file is to be associated
+     * @param content the content for the file to be associated with the specified key
+     */
+    public void putIfAbsent(String key, byte[] content) {
```

Comment:
Maybe, this function should return a boolean:
- return true, iff the file was changed
That would make implementing `putIfAbsentAndGet` easier.

Reply:
The idea behind of these method follows the `ExpiringCacheMap` implementation. That is why we have a `void putIfAbsent()` and a `byte[] putIfAbsentAndGet(...)` method.

Reply:
But then the signature should be:
```
byte[] putIfAbsentAndGet(String key, Supplier<byte[]> action)
```
and the action should only be called, if the value is absent.

Reply:
```
	public void putIfAbsent(String key, byte[] content) {
		putIfAbsentAndGet(key, () -> content);
	}
	public byte[] putIfAbsentAndGet(String key, Supplier<byte[]> action) {
		File fileInCache = getUniqueFile(key);
		if (fileInCache.exists()) {
			logger.debug("File '{}' present in cache", fileInCache.getName());
			// update time of last use
			fileInCache.setLastModified(System.currentTimeMillis());
		} else {
			writeFile(fileInCache, action.get());
		}
	}
````

Reply:
sorry, getting thing mixed up ....

Reply:
Alright. Then we are done here?